core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -149,7 +151,14 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
       nameToAddedField.put(name, newField);
     }
 
-    adds.add(newField);
+    String partitionName;
+    if (newField.name() != null) {
+      partitionName = newField.name();
+    } else {
+      partitionName = PartitionSpecVisitor.visit(schema, newField, PartitionNameGenerator.INSTANCE);
+    }
+
+    adds.add(Pair.of(newField, partitionName));
```

Comment:
Why not create a new `PartitionField` with its name set to `partitionName`? Then you wouldn't need to change as much, including the type of `adds`.

Reply:
If `name` is null, we generate default name for it at the end of `apply()` method. We need all name of added field before detect the conflict. So i change the type of `adds` to save generated name , avoid generate it twice.
if i don't change the type of `adds`, i think there are four other options:
1. just generate name twice
2. add a set method `setName(String name)` to `PartitionField`
3. generate newField again with generated name: just like
```java
if (newField.name() == null) {
    String partitionName = PartitionSpecVisitor.visit(schema, newField, PartitionNameGenerator.INSTANCE);
    newField = new PartitionField(
        newField.sourceId(), newField.fieldId(), partitionName, newField.transform());
}

adds.add(newField);
```
4. write a new static `visit` method in `PartitionSpecVisitor`, the signature is like:
```java
static <R> R visit(Schema schema, int sourceId, Transform<?, ?> transform, PartitionSpecVisitor<R> visitor)
```
Which do you think is better?

Reply:
The `PartitionField` is created in this method, as is `partitionName`. I think that by reordering a couple of statements here, you can use the correct name on the field from the start and avoid more changes.

Reply:
Regenerate `PartitionField` with `partitionName ` if `name` is null.